File: src/main/java/cat/calidos/morfeu/model/injection/ModelComponent.java
// MODEL COMPONENT . JAVA

package cat.calidos.morfeu.model.injection;

import java.net.URI;
import java.util.concurrent.ExecutionException;

import javax.inject.Named;

import com.google.common.util.concurrent.ListenableFuture;

import dagger.BindsInstance;
import dagger.producers.ProductionComponent;

import cat.calidos.morfeu.model.Model;
import cat.calidos.morfeu.problems.ValidationException;
import cat.calidos.morfeu.utils.injection.ListeningExecutorServiceModule;


/**
 * @author daniel giribet
 *///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
@ProductionComponent(modules = { ModelModule.class, ListeningExecutorServiceModule.class })
public interface ModelComponent {

ListenableFuture<Model> model() throws ValidationException, ExecutionException;

//@formatter:off
@ProductionComponent.Builder
interface Builder {
	@BindsInstance Builder identifiedBy(@Named("ModelURI") URI u);
	@BindsInstance Builder fromFetchable(@Named("FetchableModelURI") URI u);
	ModelComponent build();
}
//@formatter:on

}

/*
 * Copyright 2024 Daniel Giribet
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

File: src/main/webapp/app/model.component.ts
/*
 *    Copyright 2017 Daniel Giribet
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

import { Component, Inject, OnInit } from '@angular/core';
import { Subscription }   from 'rxjs/Subscription';


import { Model } from './model.class';
import { Widget } from './widget.class';
import { RemoteDataService } from './services/remote-data.service';

import { EventService } from './events/event.service';
import { ModelRequestEvent } from './events/model-request.event';
import { StatusEvent } from './events/status.event';


@Component({
    moduleId: module.id,
    selector: 'model',
    template: `
    <div id="model-info" class="panel panel-default" *ngIf="model">
        <div class="panel-heading">
            <h4 id="model-name" class="panel-title">Model: {{model.name}}</h4>
        </div>
        <div class="panel-body">
            <span id="model-desc">{{model.desc}}</span>
        </div>
    </div>
    `,

    styles:[`
                #model-info {}
                #model-name {}
                #model-desc {}
    `]
})

export class ModelComponent extends Widget implements OnInit {
    
model: Model;
    
    
constructor(eventService: EventService, 
            @Inject("ModelService") private modelService: RemoteDataService<Model> ) {
    super(eventService);
}
    

ngOnInit() {
    
    console.log("ModelComponent::ngOnInit()"); 
    
    this.subscribe(this.events.service.of(ModelRequestEvent).subscribe( requested =>
            this.loadModel(requested.url)           
    ));
    
}

loadModel(uri: string) {
    this.events.service.publish(new StatusEvent("Fetching model"));
    let modelURI = "/morfeu/models/"+uri;
    this.modelService.get(modelURI).subscribe( model => {
            console.log("ModelComponent::loadModel() Got model from Morfeu service ("+model.name+")");
            this.model = model;
            this.events.ok();
    },
    error => {
        this.events.problem(error);
    },
    () =>     this.events.service.publish(new StatusEvent("Fetching model", StatusEvent.DONE))
    );
}


}